File: benchmarks/table_sort.lua
-- Двойник table_sort.bsl.
--
-- Две разницы семантики, из-за которых Lua здесь заведомо в выигрыше и это
-- НЕ повод «догонять» (см. README):
--   * table.sort неустойчива, наша Сортировать устойчива и обязана такой
--     остаться — от этого зависит порядок при равных ключах;
--   * сравнение строк в Lua побайтовое, у нас — через collate (приближение
--     локали: сперва ВРег, потом исходный вид), то есть две операции
--     приведения регистра на каждое сравнение.

local ROWS = 100000

local t = {}
local seed = 1
for i = 1, ROWS do
    seed = (seed * 1103515245 + 12345) % 2147483648
    t[i] = { key = seed, name = "имя" .. i }
end

local started = os.clock()

table.sort(t, function(a, b) return a.key < b.key end)
table.sort(t, function(a, b) return a.key > b.key end)
table.sort(t, function(a, b) return a.name < b.name end)

local needle = "имя" .. ROWS
local found = nil
for i = 1, #t do
    if t[i].name == needle then
        found = t[i]
        break
    end
end

local elapsed_ms = (os.clock() - started) * 1000

print(string.format("строк: %d, нашлась: %s", #t, tostring(found ~= nil)))
print(string.format("%.3f", elapsed_ms))

File: benchmarks/table_sort.py
"""Двойник `table_sort.bsl`: три сортировки и линейный поиск."""

from __future__ import annotations

from time import perf_counter


ROWS = 100_000
table: list[list[object]] = []
seed = 1
for number in range(1, ROWS + 1):
    seed = (seed * 1_103_515_245 + 12_345) % 2_147_483_648
    table.append([seed, "имя" + str(number)])

started = perf_counter()
table.sort(key=lambda row: row[0])
table.sort(key=lambda row: row[0], reverse=True)
# `ТаблицаЗначений.Сортировать` сравнивает строки сначала без учёта
# регистра, затем в исходном виде.
table.sort(key=lambda row: (str(row[1]).upper(), row[1]))

needle = "имя" + str(ROWS)
found = next((row for row in table if row[1] == needle), None)
elapsed_ms = (perf_counter() - started) * 1_000

print(f"строк: {len(table)}, нашлась: {found is not None}")
print(f"{elapsed_ms:.3f}")
